Sentinel slicing a many pointer doesn't produce a sentinel terminated many pointer.

### Zig Version

0.12.0-dev.1631+359842f8d

### Steps to Reproduce and Observed Behavior

```zig
pub fn main() !void {
    var ptr1: [*]const u8 = undefined;
    @compileLog(ptr1[0.. :0]); // [*]const u8
    var ptr2: []const u8 = undefined;
    @compileLog(ptr2[0.. :0]); // [:0]const u8
}
```

### Expected Behavior

`ptr1[0.. :0]` should produce a `[*:0]const u8`

While there is certainly a bug in the compiler here, I think the expected behaviour in the OP is really a language proposal. It has never been possible to use this syntax to change the sentinel of a many-item pointer (at least not in the last couple of years).

If we allow adding a sentinel to the type of a many-item pointer by slicing without an end value, I believe it will be the only unsafe type-changing operation that cannot be checked (either at compile time or with a runtime safety check) and does not require a casting builtin. As far as I can tell this would introduce a new class of possible buffer overruns in safe build modes that are not currently possible without using casting builtins.

This would also have different semantics than the identical syntax for slicing an array or slice, which may make code that is generic over slices, arrays and many-item pointers more cumbersome to write. Currently you will hit a OOB safety panic or compile error when trying to add a sentinel to a slice or array as in the OP, so the fact that it is not a compile error when there is a `@compileLog` does not mean you can write valid code doing it; for example:

```zig
test {
    const d = [_]i32{ 1, 2, 3, 4, 5 };
    var slice: []const i32 = undefined;
    slice = d[0..];
    const slice_z = slice[0..: 0];
    _ = slice_z;
}
```
results in `Test [1/1] test_0... thread 8636 panic: index out of bounds: index 6, len 5`  and trying to do the sentinel-adding slice on the array results in `t.zig:1:1: error: slice end index 5 exceeds bounds of containing decl of type '[5]i32'.

Does allowing adding a sentinel via slice syntax also mean that slicing a multi-ptr slice an end value should also be able to change the sentinel of a type?
```zig
const a: [:0]const u8 = &.{1, 2, 3, 4, 5};
const b = a[0..: 7];
```
While I also think this should require a `@ptrCast`, you can at least make an O(n) scan up to the old sentinel for a safety check to make sure the element does exist. This would again have different semantics than the same syntax for slicing an array or slice as if you try to change the sentinel on a sentinel-terminated slice or array. The safety checks for slices and arrays ensure that the requested sentinel is located at `data[data.len]` and you get a sentinel mismatch safety panic or compile error:
```zig
test {
    const d = [_:0]i32{ 1, 2, 3, 4, 5 };
    _ = d[0..:7];
}
```
results in
```
t.zig:3:10: error: value in memory does not match slice sentinel
    _ = d[0..:4];
        ~^~~~~~~
t.zig:3:10: note: expected '4', found '0'
```

--- a/src/Sema.zig
+++ b/src/Sema.zig
@@ -31865,6 +31865,38 @@ fn analyzeSlice(
                 break :e try sema.coerce(block, .usize, uncasted_end, end_src);
             } else break :e try sema.coerce(block, .usize, uncasted_end_opt, end_src);
         }
+
+        // when slicing a many-item pointer, if a sentinel `S` is provided as in `ptr[a.. :S]`, it
+        // must match the sentinel of `@TypeOf(ptr)`.
+        sentinel_check: {
+            if (sentinel_opt == .none) break :sentinel_check;
+            const provided = provided: {
+                const casted = try sema.coerce(block, elem_ty, sentinel_opt, sentinel_src);
+                try checkSentinelType(sema, block, sentinel_src, elem_ty);
+                break :provided try sema.resolveConstDefinedValue(
+                    block,
+                    sentinel_src,
+                    casted,
+                    .{ .simple = .slice_sentinel },
+                );
+            };
+
+            if (ptr_sentinel) |current| {
+                if (provided.toIntern() == current.toIntern()) break :sentinel_check;
+            }
+
+            return sema.failWithOwnedErrorMsg(block, msg: {
+                const msg = try sema.errMsg(sentinel_src, "sentinel-terminated slicing of many-item pointer must match existing sentinel", .{});
+                errdefer msg.destroy(sema.gpa);
+                if (ptr_sentinel) |current| {
+                    try sema.errNote(sentinel_src, msg, "expected sentinel '{f}', found '{f}'", .{ current.fmtValue(pt), provided.fmtValue(pt) });
+                } else {
+                    try sema.errNote(ptr_src, msg, "type '{f}' does not have a sentinel", .{slice_ty.fmt(pt)});
+                }
+                try sema.errNote(src, msg, "use @ptrCast to cast pointer sentinel", .{});
+                break :msg msg;
+            });
+        }
         return sema.analyzePtrArithmetic(block, src, ptr, start, .ptr_add, ptr_src, start_src);
     };
 

--- a/test/behavior/slice.zig
+++ b/test/behavior/slice.zig
@@ -628,9 +628,6 @@ test "slice syntax resulting in pointer-to-array" {
             comptime assert(@TypeOf(ptr[1..][0..2]) == *[2]u8);
             comptime assert(@TypeOf(ptr[1..][0..4]) == *[4]u8);
             comptime assert(@TypeOf(ptr[1..][0..2 :4]) == *[2:4]u8);
-            comptime assert(@TypeOf(ptr[1.. :0][0..2]) == *[2]u8);
-            comptime assert(@TypeOf(ptr[1.. :0][0..4]) == *[4]u8);
-            comptime assert(@TypeOf(ptr[1.. :0][0..2 :4]) == *[2:4]u8);
 
             var ptr_z: [*:0]u8 = &array;
             comptime assert(@TypeOf(ptr_z[1..][0..2]) == *[2]u8);

--- a/test/cases/compile_errors/slice_of_many-item_pointer_preserves_sentinel.zig
+++ b/test/cases/compile_errors/slice_of_many-item_pointer_preserves_sentinel.zig
@@ -1,0 +1,18 @@
+comptime {
+    var ptr: [*]const u8 = undefined;
+    _ = ptr[0.. :0];
+}
+
+comptime {
+    var ptrz: [*:0]const u8 = undefined;
+    _ = ptrz[0.. :1];
+}
+
+// error
+//
+// :3:18: error: sentinel-terminated slicing of many-item pointer must match existing sentinel
+// :3:9: note: type '[*]const u8' does not have a sentinel
+// :3:12: note: use @ptrCast to cast pointer sentinel
+// :8:19: error: sentinel-terminated slicing of many-item pointer must match existing sentinel
+// :8:19: note: expected sentinel '0', found '1'
+// :8:13: note: use @ptrCast to cast pointer sentinel

